ui-admin/src/util/tableUtils.tsx

```diff
@@ -266,6 +269,73 @@ export function ColumnVisibilityControl<T>({ table }: {table: Table<T>}) {
   </div>
 }
 
+/**
+ * Converts a cell value to an escaped string for csv export
+ */
+function cellToCsvString(cellType: string, cellValue: unknown): string {
+  switch (cellType) {
+    case 'string':
+      return cellValue ? escapeCsvValue(cellValue as string) : ''
+    case 'instant':
+      return escapeCsvValue(instantToDefaultString(cellValue as number))
+    case 'boolean':
+      return cellValue as boolean ? 'true' : 'false'
+    default:
+      return cellValue ? escapeCsvValue(cellValue as string) : ''
+  }
+}
+
+/**
+ * Adds a control to download the table data as a csv file
+ */
+export function DownloadControl<T>({ table, fileName }: {table: Table<T>, fileName: string}) {
+  const [show, setShow] = useState(false)
+
+  const download = () => {
+    const headers = table.getFlatHeaders().map(header => {
+      return header.id
+    }).join(',')
+
+    const rows = table.getFilteredRowModel().rows.map(row => {
+      return row.getVisibleCells().map(cell => {
+        const cellType = cell.column.columnDef.meta?.columnType || 'string'
+        return cellToCsvString(cellType, cell.getValue())
+      }).join(',')
+    }).join('\n')
+
+    const blob = new Blob([headers, '\n', rows], { type: 'text/plain' })
+    saveBlobAsDownload(blob, `${fileName}.csv`)
+  }
+
+  return <div className="ms-auto">
+    <button className="btn btn-light border m-1" disabled={isEmpty(table.getFilteredRowModel().rows)}
+      onClick={() => setShow(!show)} aria-label="download table data">
+      <FontAwesomeIcon icon={faDownload} className="fa-lg"/> Download
+    </button>
+    { show && <Modal show={show} onHide={() => setShow(false)}>
+      <Modal.Header closeButton>
+        <Modal.Title>
+                Download
+        </Modal.Title>
+      </Modal.Header>
+      <Modal.Body>
+        <div className="border-b border-black">
+                This will download <strong>{table.getFilteredRowModel().rows.length}</strong> rows
+                to <code>{fileName}.csv</code>. Your current row and column filters will be applied to the
+                downloaded data.
```

Comment:
possible wordsmithing for more conciseness? 
```suggestion
                Download <strong>{table.getFilteredRowModel().rows.length}</strong> rows
                to <code>{fileName}.csv</code>. The current data filters and shown columns will be applied.
```